crates/ruff_python_parser/src/lexer.rs

```diff
@@ -579,6 +728,28 @@ impl<'source> Lexer<'source> {
     // This is the main entry point. Call this function to retrieve the next token.
     // This function is used by the iterator implementation.
     pub fn next_token(&mut self) -> LexResult {
+        if let Some(fstring_context) = self.fstring_stack.last() {
+            if !fstring_context.is_in_expression()
+                // Avoid lexing f-string middle/end if we're sure that this is
+                // the start of a f-string expression i.e., `f"{foo}"` and not
+                // `f"{{foo}}"`.
+                && (self.cursor.first() != '{' || self.cursor.second() == '{')
+                // Avoid lexing f-string middle/end if we're sure that this is
+                // the end of a f-string expression. This is only for when
+                // the `}` is after the format specifier i.e., `f"{foo:.3f}"`
+                // because the `.3f` is lexed as `FStringMiddle` and thus not
+                // in a f-string expression.
+                && (!fstring_context.is_in_format_spec() || self.cursor.first() != '}')
+            {
```

Comment:
I don't really like this that much. I think we're sacrificing readability a bit. Without this, the `lex_fstring_middle_or_end` function would return `Result<Option<Tok>, ...>` and this condition will be identified initially in the function loop itself.

Reply:
I've reverted this change although I'm open to suggestions.